### ISSUE TITLE
Auto project columns based on `AsQueryable<T>` and `AsEnumerable<T>` generic type

### ISSUE THREAD

**Issue by BrunoBlanes** [dotnet/efcore#29878]
## The problem

Suppose I have an API and want to query the following `Foo` entity by its `Name` property:
```c#
public sealed class Foo : IEntityBase, IQueryableEntity
{
    public Foo()
    {
	Id = Guid.NewGuid();
	Name = string.Empty;
	Created = DateTimeOffset.UtcNow;
    }

    public Guid Id { get; init; }
    public string Name { get; init; }
    public DateTimeOffset Created { get; init; }
}

public interface IEntityBase
{
    Guid Id { get; }
    DateTimeOffset Created { get; }
}

public interface IQueryableEntity
{
    Guid Id { get; }
    string Name { get; }
}
```

This is what the request method would look like:
```c#
public async Task<IQueryableEntity[]> OnGetAsync(string query, CancellationToken cancellationToken)
{
    IQueryableEntity[] result = await context.Foos.AsQueryable<IQueryableEntity>()
	.Where(x => x.Name.Contains(query))
	.ToArrayAsync(cancellationToken);
    return result;
}
```

However the generated SQL includes all columns:
```sql
SELECT [c].[Id], [c].[Created], [c].[Name]
FROM [dbo].[Foos] AS [c]
WHERE (@__query_0 LIKE N'') OR CHARINDEX(@__query_0, [c].[Name]) > 0
```

The [documentation says](https://learn.microsoft.com/en-us/ef/core/performance/efficient-querying#project-only-properties-you-need) I should use `.Select()` to project only the necessary columns. So that would look like this:
```c#
    IQueryableEntity[] result = await context.Foos.AsQueryable<IQueryableEntity>()
	.Where(x => x.Name.Contains(query))
	.Select(x => new Foo
	{
	    Id = x.Id,
	    Name = x.Name
	}).ToArrayAsync(cancellationToken);
```
Although this *does* work, it no longer returns an `IQueryableEntity` array, instead it returns `Foo[]`. The problem with that is that if you return it, ASP.NET Core will include the `Created` property in the JSON result because it was initialized in the constructor, and this is `Foo` now, no longer `IQueryableEntity`.

Of course using an anonymous type in `.Select()` would fix this problem, but I can no longer work with the result as an `IQueryableEntity` and seems like there's a lot more casting involved than need be (I could be *and probably am* wrong here).

## The proposal
Take advantage of the `AsQueryable<T>` and `AsEnumerable<T>` and use `T` as the type for projection. This would probably be a breaking change if made default behavior, so instead introduce new extension methods for `.Select()`:
```c#
public static IQueryable<T> Select<T>(this IQueryable<T> source);
public static IEnumerable<T> Select<T>(this IEnumerable<T> source);
```

**Comment by roji** [comment]
> Although this does work, it no longer returns an IQueryableEntity array, instead it returns Foo[].

Since arrays are covariant, why not just cast the resulting `Foo[]` into an `IQueryableEntity[]`?

Another possibility is to cast Foo up to IQueryableEntity inside your projection:

```c#
	.Select(x => (IQueryableEntity)new Foo
	{
	    Id = x.Id,
	    Name = x.Name
	}).ToArrayAsync(cancellationToken);
```

**Comment by BrunoBlanes** [comment]
> Since arrays are covariant, why not just cast the resulting `Foo[]` into an `IQueryableEntity[]`?

Because of casting. Although sure, both of your suggestions would fix this issue, they look and feel cumbersome. Having this be native behavior would be more in line with what the expected behavior should be.

If I am querying an `IQueryableEntity` then I don't care about its other properties in my code, why should EF query those from the database anyway?

**Comment by roji** [comment]
First, in your suggestion above:

```c#
public static IQueryable<T> Select<T>(this IQueryable<T> source);
public static IEnumerable<T> Select<T>(this IQueryable<T> source);
```

These methods need to have two generic parameters, not one - right? One for the source (Foo) and one for the projection (IQueryableEntity), right?

In any case, this already exists - it's the [Cast](https://learn.microsoft.com/en-us/dotnet/api/system.linq.queryable.cast?view=net-7.0#system-linq-queryable-cast-1(system-linq-iqueryable)) LINQ operator.

Now, EF Core's implementation of this operator doesn't fetch only the properties for the projected type. I'm not convinced that would be good behavior; for one thing, users may check the actual type of the entity, and then cast it back down to access specific properties which aren't part of the interface.

More generally, the goal when implementing the queryable operators is to approximate the behavior of the enumerable ones as much as possible. Dropping properties because they're not part of the interface goes against that, since with the enumerable implementation they're there (it just does a cast). That's why it makes sense for users to explicitly say when they want to drop properties, rather than assume that a projection to an interface implies it.

**Comment by BrunoBlanes** [comment]
> First, in your suggestion above:
> 
> ```cs
> public static IQueryable<T> Select<T>(this IQueryable<T> source);
> public static IEnumerable<T> Select<T>(this IQueryable<T> source);
> ```
> 
> These methods need to have two generic parameters, not one - right? One for the source (Foo) and one for the projection (IQueryableEntity), right?

I was assuming that you would use the source as the projection parameter, therefore I think one would suffice, so the following could be achieved:

```c#
IQueryableEntity[] result = await context.Foos.AsQueryable<IQueryableEntity>()
	.Where(x => x.Name.Contains(query))
	.Select().ToArrayAsync(cancellationToken);
```

> In any case, this already exists - it's the [Cast](https://learn.microsoft.com/en-us/dotnet/api/system.linq.queryable.cast?view=net-7.0#system-linq-queryable-cast-1(system-linq-iqueryable)) LINQ operator.

Yes, of course, I meant it in a way to simplify it, so you don't need to do any explicit casting.

> Dropping properties because they're not part of the interface goes against that, since with the enumerable implementation they're there (it just does a cast). That's why it makes sense for users to explicitly say when they want to drop properties, rather than assume that a projection to an interface implies it.

I now understand what the goal of the implementation is, and I know some people may want to cast it back to the original type and work on those properties, hence why I thought this would be a breaking change if made default behavior.

My suggestion was mainly the new parameterless `.Select()` method. Since the general idea of this method is to do database projection, it wouldn't go against it and wouldn't be a breaking change.

**Comment by roji** [comment]
>> In any case, this already exists - it's the [Cast](https://learn.microsoft.com/en-us/dotnet/api/system.linq.queryable.cast?view=net-7.0#system-linq-queryable-cast-1(system-linq-iqueryable)) LINQ operator.
>
> Yes, of course, I meant it in a way to simplify it, so you don't need to do any explicit casting.

A new parameter-less operator doesn't seem any less explicit than the existing Cast operator.

> My suggestion was mainly the new parameterless .Select() method. Since the general idea of this method is to do database projection, it wouldn't go against it and wouldn't be a breaking change.

The point is that the Queryable version of the new Select operator which you're proposing would be inconsistent with the Enumerable version. The Enumerable version would simply be a no-op (or at most an up-cast), returning the same instances it gets as a parameter (with all their properties). EF's implementation of the Queryable version, on the other hand, would exclude properties no included in the (interface) type, which would prevent down-casting back. So your two proposed new operators would be inconsistent.

Also, as I wrote above, I'm still skeptical that dropping the properties is the right implicit behavior for your proposed Queryable operator. Sure, since it's new, there's not going to be any breaking change; but that doesn't mean any behavior there makes sense. There's nothing on a parameter-less Select operator that indicates that some properties are dropped.

Finally, the enumerable version of your proposed operator would generally be useless - it's hard to imagine a case where it would be useful.

**Comment by BrunoBlanes** [comment]
> The point is that the Queryable version of the new Select operator which you're proposing would be inconsistent with the Enumerable version. The Enumerable version would simply be a no-op (or at most an up-cast), returning the same instances it gets as a parameter (with all their properties). EF's implementation of the Queryable version, on the other hand, would exclude properties no included in the (interface) type, which would prevent down-casting back. So your two proposed new operators would be inconsistent.

I'm sorry, I don't see how they'd be different or inconsistent. Why would the `IEnumerable<T>` version be a simple cast operation?

> Also, as I wrote above, I'm still skeptical that dropping the properties is the right implicit behavior for your proposed Queryable operator.

Isn't the point of `.Select()` to be able to project specific columns?

**Comment by roji** [comment]
> Why/how would the IEnumerable<T> version be a simple cast operation?

You proposed adding the following operator:

> public static IEnumerable<T> Select<T>(this IQueryable<T> source);

What do you propose as the implementation for this (the **enumerable** version)? What could it possibly do other than cast to T?

>> Also, as I wrote above, I'm still skeptical that dropping the properties is the right implicit behavior for your proposed Queryable operator.
>
> Isn't the point of .Select() to be able to project only specific columns?

Not necessarily, no. Keep in mind that Select exists also as an enumerable operator, not just as an EF queryable provider. Select in general is about projecting, which could mean narrowing down, or possibly even adding other properties. There's certainly nothing in the idea of projecting to some type T that implies that all properties that aren't a part of T should be removed.

In other words, in the EF context, Select is indeed frequently (but not exclusively) used to reduce the set of columns fetched from the database. But that isn't anything that's inherent in the concept of Select, even across different queryable LINQ providers.

**Comment by roji** [comment]
> Maybe you'd in fact need two type parameters for this to work, since you'd need to check if the source can be casted.

You do need two type parameters, simply because two types are involved: the input type (Foo in your example) and the output type (IQueryableEntity).

I think the point you are missing, is that your proposed enumerable and queryable operators wouldn't be consistent between themselves, since the enumerable one would simply cast; this means that the user will still be able to cast back down to the real type and access properties which aren't on the projected type. This wouldn't be true on the queryable version.

In addition, the enumerable version is, on its own, useless - it's identical to the already-existing Cast operator. It's hard to make a case for introducing a general LINQ operator which only makes sense in the EF queryable context.

**Comment by BrunoBlanes** [comment]
> In addition, the enumerable version is, on its own, useless - it's identical to the already-existing Cast operator.

Except for projection, right? Although it would be identical to the `Cast` operator in the sense that it would do the same, it would also project only specific columns, something that `Cast` doesn't do, or does it and I've missed that in the docs?

**Comment by ilmax** [comment]
A bit off topic maybe, but in order to better control serialization, I always copy what I read from the db to some sort of DTO. 
If you want to directly serialize the ef entities you will go down a painful path because more often than not, what makes sense over the wire may not be the best way to store data in a db and viceversa.

**Comment by roji** [comment]
>> In addition, the enumerable version is, on its own, useless - it's identical to the already-existing Cast operator.
>
> Except for projection, right? Although it would be identical to the Cast operator in the sense that it would do the same, it would also project only specific columns, something that Cast doesn't do, or does it and I've missed that in the docs?

Not the enumerable version. Once again, as I've written above, what is your proposed **implementation** for your proposed enumerable Select operator? What would it possibly do that's different from the Cast operator (note that I'm talking about the **enumerable** operator, not the **queryable** one).

> A bit off topic maybe, but in order to better control serialization, I always copy what I read from the db to some sort of DTO.

That's certainly possible, but it effectively means never using EF's change tracking capabilities. It's true that it's always good to be aware of exactly which columns you're projecting out of a database, but that in itself isn't necessarily a reason to never just query entities out when that makes sense.

**Comment by ilmax** [comment]
> That's certainly possible, but it effectively means never using EF's change tracking capabilities.

If you're in a reading scenario, change tracking may actually be turned off on the whole context to improve perf.

My suggestion to decouple entities and DTOs was due to this sentence in the issue:

> The problem with that is that if you return it, ASP.NET Core will include the Created property in the JSON result because it was initialized in the constructor, and this is Foo now, no longer IQueryableEntity.

So my suggestion was to use two classes for two different purposes: the entity to store and read data from the dB and the DTO to fine tune what get serialized over the wire. 
Using this approach doesn't mean giving up control on what/how we read from the database, in fact we can do something like the following:

```csharp
IQueryableEntity[] result = await context.Foos.AsQueryable<IQueryableEntity>()
	.Where(x => x.Name.Contains(query))
	.Select(x => new EntityDto
	{
	    Id = x.Id,
	    Name = x.Name
	}).ToArrayAsync(cancellationToken);
```
so we only read what we need.

This is just my 2cent because I had similar problems in the past and as soon as I decided to introduce a separate DTO, I had no more issues on that front.

Lastly using a DTOs is not free of charge, you have several costs involved with using them:
- You have to write and maintain more core
- You will incur a performance penalty if you go through a mapping process
- Probably more
But it solves some other issues nicely.

**Comment by roji** [comment]
>>> A bit off topic maybe, but in order to better control serialization, I always copy what I read from the db to some sort of DTO.
>> That's certainly possible, but it effectively means never using EF's change tracking capabilities.
> If you're in a reading scenario, change tracking may actually be turned off on the whole context to improve perf.

It's one thing to turn off change tracking in specific perf-sensitive areas, but quite another to systematically always project out to a DTO.

But I'm not disagreeing with the general concept... It's true that LINQ tends to encourage over-fetching by getting all entity properties by default, requiring a specific gesture to selectively project only a subset. I agree that in various scenarios (especially disconnected ones where change tracking is frequently less relevant) systematic projection can be a good idea.

One part where I'm slightly less clear about, is the need for a separate DTO type - you can usually just project out the same type you're reading from the database, but populating only the properties you need (you did this in your original post above):

```c#
return (IBlogView[])context.Blogs.Select(b => new Blog { Name = b.name, Rating = b.rating }).ToArray();
```

You're still not exposing Blog directly - only through IBlogView which functions a bit like your DTO - but without having to copy data across (i.e. just expose a view rather than a copy). Of course, if you're just sending data back via JSON (e.g. a typical web API), anonymous objects can make this even simpler, removing the need for even a view interface (since they're serialized to JSON just as well, no need for a special named type).

To summarize, I definitely agree that it's a good idea to carefully think about which properties you really need to get from the database and expose out. EF provides several techniques to do this (project out to the same type + view interface, project out to an anonymous type).

I agree it may be interesting think about making this even easier via a dedicated terminating operator effectively projects out to an untracked instance, populating only the properties which exist on an interface implemented by that type. I don't think it makes sense for that operator to be simply called Select (since it has some specific semantics which Select does not have), and I also don't think an enumerable version of that operator makes sense (it would rather be an EF-specific thing).

**Comment by ilmax** [comment]
> I agree that in various scenarios (especially disconnected ones where change tracking is frequently less relevant) systematic projection can be a good idea.

Excellent point, I'm presuming we're talking about a disconnected scenario here due to the mention of asp.net core serialization

> One part where I'm slightly less clear about, is the need for a separate DTO type - you can usually just project out the same type you're reading from the database, but populating only the properties you need (you did this in your original post above):
```csharp
return (IBlogView[])context.Blogs.Select(b => new Blog { Name = b.name, Rating = b.rating }).ToArray();
```

Sure you can do that, but when the result of that query will be serialized to json and sent over the wire, it will include all the other properties of the Blog entity as well with a null value (less than ideal).
You can probably tweak the json serialization to skip null properties but this only allows you to get away in a simple case.
Additionally it may be confusing to return a partially initialized entity, imagine for example if the client code starts accessing the `Posts` navigation property after the above query it will get a nullref.

Also whenever you need to tweak the json serialization more, you will soon run into other issues, that's why I suggested to separate out the DTO to have greater control over serialization.

**Comment by roji** [comment]
> Sure you can do that, but when the result of that query will be serialized to json and sent over the wire, it will include all the other properties of the Blog entity as well with a null value (less than ideal).

Right, but if you're in a JSON serialization scenario then anonymous types are perfect, no? The "view interface" pattern was meant for programmatic APIs where you need to return a strongly typed .NET type to the user (and anonymous types aren't an option).

> Additionally it may be confusing to return a partially initialized entity, imagine for example if the client code starts accessing the Posts navigation property after the above query it will get a nullref.

The point (for programmatic API access) is to not expose stuff you're not going to populate - that's what IBlogView is there for (it wouldn't contain a Posts property). Sure, the user can technically downcast the interface back to Blog, but that's generally not something that should be guarded against.

So I'm still not seeing why a separate DTO type is needed, as opposed to a smiple view interface (for programmatic APIs) or anonymous types (for JSON serialization).

**Comment by ilmax** [comment]
> Right, but if you're in a JSON serialization scenario then anonymous types are perfect, no? The "view interface" pattern was meant for programmatic APIs where you need to return a strongly typed .NET type to the user (and anonymous types aren't an option).

Well, as it's always the case it depends :) Imagine you are querying different things in different data stores and then aggregate the result to return it to the client, in that case anonymous types are a no go.

> The point (for programmatic API access) is to not expose stuff you're not going to populate - that's what IBlogView is there for (it wouldn't contain a Posts property). Sure, the user can technically downcast the interface back to Blog, but that's generally not something that should be guarded against

Hiding stuff behind a `IBlogView` may be enough to ensure client code is discouraged to do the wrong thing (e.g. accessing Posts collection), but when you try to serialize IBlogView[] that's actually a Blog[] you won't get the expected json serialized by asp.net core. 

Just to double check this I made a sample [here](https://github.com/ilmax/BlogViewExample) and, unless I'm missing something, when you hit the /efquery and search for e.g. "blog" you get back more than what the `IBlogView` exposes.

**Comment by roji** [comment]
> Hiding stuff behind a IBlogView may be enough to ensure client code is discouraged to do the wrong thing (e.g. accessing Posts collection), but when you try to serialize IBlogView[] that's actually a Blog[] you won't get the expected json serialized by asp.net core.

My point was that for serialization you can just use an anonymous type instead.

But yeah, if you're writing a strongly-typed API whose results later get serialized, that's true.

**Comment by ajcvickers** [comment]
Note from triage: we discussed this but we don't plan to make any changes in this area. You might want to consider AutoMapper for reducing the code needed for simple, by convention, projections.
